btree/btree.c

```diff
@@ -3721,35 +3710,46 @@ int64_t btree_iter_tick(struct m0_sm_op *smop)
 							       oi->i_key_found);
 				/**
 				 * In the following cases jump to LOCK state:
-				 * 1. Normal pass without sibling load flag.
-				 * 2. Sibling load flag and found record idx
-				 * is less than number of records in leaf node.
-				 * 3. It is the rightmost record in the btree.
+				 * 1. the found key idx is within the valid
+				 *    index range of the node.
+				 * 2.i_pivot is equal to -1. It means, tree
+				 *   traversal reached at the leaf level without
+				 *   finding any valid sibling in the non-leaf
+				 *   levels.
+				 *   This indicates that the search key is the
+				 *   boundary key (rightmost for NEXT flag and
+				 *   leftmost for PREV flag).
 				 */
-				if (!(bop->bo_flags & BOF_SIBLING) ||
-				    (index_is_valid(lev)) ||
-				    (oi->i_pivot == -1))
+				if (index_is_valid(lev) || oi->i_pivot == -1)
 					return P_LOCK;
 
 				/**
-				 * Start traversing the sibling node path.
-				 * If the pivot node is still valid, l
-				 * load sibling idx's child node at pivot level
-				 * else clean up and restart state machine.
+				 * We are here, it means we want to load
+				 * sibling node of the leaf node.
+				 * Start traversing the sibling node path
+				 * starting from the pivot level. If the node
+				 * at pivot level is still valid, load sibling
+				 * idx's child node else clean up and restart
+				 * state machine.
 				 */
 				lev = &oi->i_level[oi->i_pivot];
 
-				if (node_is_valid(lev->l_node) != 0) {
+				if (!node_is_valid(lev->l_node)) {
 					node_op_fini(&oi->i_nop);
+					bop->bo_flags |= BOF_LOCKALL;
 					return m0_sm_op_sub(&bop->bo_op,
 							    P_CLEANUP, P_SETUP);
 				}
-				if (lev->l_seq != lev->l_node->n_seq)
+				if (lev->l_seq != lev->l_node->n_seq) {
+					bop->bo_flags |= BOF_LOCKALL;
 					return m0_sm_op_sub(&bop->bo_op,
 							    P_CLEANUP, P_SETUP);
+				}
+
 				s.s_node = lev->l_node;
 				s.s_idx = sibling_index_get(lev->l_idx,
-							    bop->bo_flags, true);
+							    bop->bo_flags,
+							    true);
```

Comment:
After this line you want to reconfirm if this index is valid because we are in unlocked state and some other thread might have deleted the next index.

Reply:
We are checking the node validity(line 3737) and its sequence number (line 3743) before getting the next index. If there is any change in this node (index deletion or complete node deletion), then we will start the state machine again in LOCKED state.

Reply:
I am still worried about this. Let us keep the code the way it is for now, just write a comment here with a question if index validity should be checked.

Reply:
Ok.